### PR TITLE
Use GroovySourceFileAllowlist to adapt to SECURITY-359 changes

### PR DESCRIPTION
**Pull request by dwnusbaum**
See https://github.com/jenkinsci/workflow-cps-plugin/pull/538. The fix for SECURITY-359 added a new API that should be used by plugins that define Groovy DSLs. Declarative defines two main kinds of Groovy DSLs:

* `ModelInterpreter.groovy` via `ModelStepLoader`: The main declarative syntax
* Various Groovy files via subclasses of `WithScriptDescribable`: Extensions related to `agent` (`DeclarativeAgent`, used by other plugins, specifically `docker-workflow`, `kubernetes`, and `amazon-ecs`) and `when` (`DeclarativeStageConditional`).`

This PR implements `GroovySourceFileAllowlist` for various individual files, but also for `WithScriptDescribable` in a generic way that automatically allows the scripts used by subclasses. This allows us to remove entries from `default-allowlist` upstream in `workflow-cps`.

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: pipeline-model-extensions/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/withscript/WithScriptDescribable.java
Comment (dwnusbaum) [inline pipeline-model-extensions/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/withscript/WithScriptDescribable.java:70]: Strictly speaking, we don't have to remove this fallback. However, it is obsolete as far as I can tell, and the fallback interacts weirdly with SECURITY-359, since it causes any files that are rejected by the allowlist to be loaded using a regular Java class loader and then parsed it using the _untrusted_ `GroovyClassLoader` for the Pipeline, causing the Groovy source file to be sandbox-transformed, which leads to unexpected `RejectedAccessException`s and confusing error messages.

File: pipeline-model-extensions/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/withscript/WithScriptDescriptor.java
Comment (dwnusbaum) [inline pipeline-model-extensions/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/withscript/WithScriptDescriptor.java:106]: This implementation of `GroovySourceFileAllowlist` is used for implementations of `DeclarativeAgent` and `DeclarativeStageConditional` (aka `when` conditions). `registerScript` is called by the `WithScriptDescriptor` constructor, so it automatically handles most implementations (barring a few special cases like abstract Groovy scripts), even those in other plugins like `docker-workflow` and `kubernetes`.

File: pipeline-model-extensions/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/agent/DeclarativeAgent.java
Comment (dwnusbaum) [inline pipeline-model-extensions/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/agent/DeclarativeAgent.java:111]: We need this because `DeclarativeAgent` itself is abstract. Another option could be to hard-code this file in `WithScriptDescriptor`.

File: pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/ModelStepLoader.java
Comment (jglick) [inline pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/ModelStepLoader.java:79]:
```suggestion
        private final String scriptUrl = ModelInterpreterAllowlist.class.getResource("/org/jenkinsci/plugins/pipeline/modeldefinition/ModelInterpreter.groovy").toString();
```

and fix imports and similarly for others:

* do not risk using the wrong `ClassLoader` if subclassed
* no need for `getClassLoader` call if you give an absolute path
* `requireNonNull` redundant since `toString` would fail anyway if the resource were missing

**Comment by dwnusbaum** [comment]
Rebuilding

File: pipeline-model-extensions/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/withscript/WithScriptDescriptor.java
Comment (jglick) [inline pipeline-model-extensions/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/withscript/WithScriptDescriptor.java:50]: SpotBugs might complain that you are calling an overridable method from a constructor?

File: pom.xml
Comment (jglick) [inline pom.xml:106]: This override is just here temporarily right?

**Review by jglick — APPROVED** [review]

File: pom.xml
Comment (dwnusbaum) [reply pom.xml:106]: I have it set to this version right now to test that everything works once the relevant entries are removed from `default-allowlist` upstream. Once this PR has been reviewed, I will change it to `2692.v76b_089ccd026` (first version with the new API) so it can be merged and released, unless that version of `workflow-cps` has made it into a BOM release in the meantime, in which case I will update the BOM and delete this.

File: pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/ModelStepLoader.java
Comment (dwnusbaum) [reply pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/ModelStepLoader.java]: All good points, I made the corresponding changes in [e1feeac](https://github.com/jenkinsci/pipeline-model-definition-plugin/pull/529/commits/e1feeac84e37eaa879c04da5e21363f509ba707d).

File: pipeline-model-extensions/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/withscript/WithScriptDescriptor.java
Comment (dwnusbaum) [reply pipeline-model-extensions/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/withscript/WithScriptDescriptor.java:50]: Yeah, I was surprised that it does not complain about this. Maybe it is because `getScriptResource` is `protected`, so SpotBugs assumes the behavior is intentional.

**Comment by dwnusbaum** [comment]
I am requesting reviews but leaving this in draft state so that it is not accidentally merged with the incremental dependencies.

**Review by car-roll — APPROVED** [review]

**Comment by jglick** [comment]
Is there something holding up release of this? I was just working on something in #533 that would have added another file next to `CheckoutScript.groovy` and found that it is not yet possible.

File: pom.xml
Comment (dwnusbaum) [reply pom.xml]: No BOM release yet, so I downgraded `workflow-cps` in [36347d3](https://github.com/jenkinsci/pipeline-model-definition-plugin/pull/529/commits/36347d379ba56dc60bacb17e7c3f9f6ab4eadfbc).

**Comment by jglick** [comment]
No rush, I can work around it for now.

**Review by yaroslavafenkin — APPROVED** [review]

File: pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/when/impl/ChangeLogConditional.java
Comment (dwnusbaum) [reply pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/when/impl/ChangeLogConditional.java]: Ah, yeah this did not work with `ClassLoader.getResource` but now it is ok. Done in [778707d](https://github.com/jenkinsci/pipeline-model-definition-plugin/pull/529/commits/778707d02aaac703b74cca47751fd5976dae44f3).

**Review by jglick — APPROVED** [review]

**Comment by jglick** [comment]
Or just enable automerge, if you have admin rights to the repo.

**Pull request merged**
